ql/src/java/org/apache/hadoop/hive/ql/optimizer/TopNKeyPushdownProcessor.java

```diff
@@ -0,0 +1,400 @@
+import org.apache.hadoop.hive.ql.exec.Operator;
+import org.apache.hadoop.hive.ql.exec.ReduceSinkOperator;
+import org.apache.hadoop.hive.ql.exec.SelectOperator;
+import org.apache.hadoop.hive.ql.exec.TopNKeyOperator;
+import org.apache.hadoop.hive.ql.lib.Node;
+import org.apache.hadoop.hive.ql.lib.NodeProcessor;
+import org.apache.hadoop.hive.ql.lib.NodeProcessorCtx;
+import org.apache.hadoop.hive.ql.parse.SemanticException;
+import org.apache.hadoop.hive.ql.plan.ExprNodeDesc;
+import org.apache.hadoop.hive.ql.plan.ExprNodeDescUtils;
+import org.apache.hadoop.hive.ql.plan.GroupByDesc;
+import org.apache.hadoop.hive.ql.plan.JoinCondDesc;
+import org.apache.hadoop.hive.ql.plan.JoinDesc;
+import org.apache.hadoop.hive.ql.plan.OperatorDesc;
+import org.apache.hadoop.hive.ql.plan.ReduceSinkDesc;
+import org.apache.hadoop.hive.ql.plan.TopNKeyDesc;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Stack;
+
+import static org.apache.hadoop.hive.ql.optimizer.TopNKeyProcessor.copyDown;
+
+public class TopNKeyPushdownProcessor implements NodeProcessor {
+  private static final Logger LOG = LoggerFactory.getLogger(TopNKeyPushdownProcessor.class);
+
+  @Override
+  public Object process(Node nd, Stack<Node> stack, NodeProcessorCtx procCtx,
+      Object... nodeOutputs) throws SemanticException {
+    pushdown((TopNKeyOperator) nd);
+    return null;
+  }
+
+  private void pushdown(TopNKeyOperator topNKey) throws SemanticException {
+
+    final Operator<? extends OperatorDesc> parent =
+        topNKey.getParentOperators().get(0);
+
+    switch (parent.getType()) {
+      case SELECT:
+        pushdownThroughSelect(topNKey);
+        break;
+
+      case FORWARD:
+        moveDown(topNKey);
+        pushdown(topNKey);
+        break;
+
+      case GROUPBY:
+        pushdownThroughGroupBy(topNKey);
+        break;
+
+      case REDUCESINK:
+        pushdownThroughReduceSink(topNKey);
+        break;
+
+      case MERGEJOIN:
+      case JOIN:
+        {
+          final CommonJoinOperator<? extends JoinDesc> join =
+              (CommonJoinOperator<? extends JoinDesc>) parent;
+          final JoinCondDesc[] joinConds = join.getConf().getConds();
+          final JoinCondDesc firstJoinCond = joinConds[0];
+          for (JoinCondDesc joinCond : joinConds) {
+            if (!firstJoinCond.equals(joinCond)) {
+              return;
+            }
+          }
+          switch (firstJoinCond.getType()) {
+            case JoinDesc.FULL_OUTER_JOIN:
+              pushdownThroughFullOuterJoin(topNKey);
+              break;
+
+            case JoinDesc.LEFT_OUTER_JOIN:
+              pushdownThroughLeftOuterJoin(topNKey);
+              break;
+
+            case JoinDesc.RIGHT_OUTER_JOIN:
+              pushdownThroughRightOuterJoin(topNKey);
+              break;
+
+            case JoinDesc.INNER_JOIN:
```

Comment:
We cannot pushdown through inner Join since you have no guarantees that a certain inner record is actually producing an output record.